### PR TITLE
Update ultralytics models (take two)

### PR DESCRIPTION
**Pull request by dreadatour**
Update Ultralytics YOLO models to be used in SaaS.

This adds:
- `Segment` model and common `YoloBBox`, `YoloBBoxes`, `YoloPose`, `YoloPoses`, `YoloSegment` and `YoloSegments` models to be used to parse ultralytics results
- Ultralytics models usage examples
- Move models from `datachain.lib.models` to `datachain.model` (I suppose to move [datachain.lib.data_model](https://github.com/iterative/datachain/blob/main/src/datachain/lib/data_model.py) to `datachain.model` in future to slowly get rid of `lib` package)

### CODE REVIEW

File: src/datachain/model/bbox.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    coords: list[int] = Field(default=None)
+
+    @staticmethod
+    def from_list(coords: list[float], title: str = "") -> "OBBox":
+        assert (
+            len(coords) == 8
+        ), "Oriented bounding box coordinates must be a list of 8 floats."
+        assert all(
+            isinstance(value, (int, float)) for value in coords
+        ), "Oriented bounding box coordinates must be integers or floats."
```
Comment (mattseddon) [inline src/datachain/model/bbox.py:83]: [N] above assert states floats only

File: src/datachain/model/ultralytics/bbox.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
+"""
+This module contains the YOLO models.
```
Comment (mattseddon) [inline src/datachain/model/ultralytics/bbox.py:2]: [Q] Does this need to be repeated in every file? Should this refer to the model in the file?

File: src/datachain/model/bbox.py (hunk 1 above)
Comment (mattseddon) [reply src/datachain/model/bbox.py:83]: same for BBox model

**Review by mattseddon — APPROVED** [review]

**Comment by dreadatour** [comment]
> Attention: Patch coverage is `73.33333%` with `64 lines` in your changes missing coverage. Please review.
> 
> [src/datachain/model/ultralytics/bbox.py](https://app.codecov.io/gh/iterative/datachain/pull/613?src=pr&el=tree&filepath=src%2Fdatachain%2Fmodel%2Fultralytics%2Fbbox.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative#diff-c3JjL2RhdGFjaGFpbi9tb2RlbC91bHRyYWx5dGljcy9iYm94LnB5)	51.61%	[30 Missing ⚠️ ](https://app.codecov.io/gh/iterative/datachain/pull/613?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative)
> [src/datachain/model/ultralytics/pose.py](https://app.codecov.io/gh/iterative/datachain/pull/613?src=pr&el=tree&filepath=src%2Fdatachain%2Fmodel%2Fultralytics%2Fpose.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative#diff-c3JjL2RhdGFjaGFpbi9tb2RlbC91bHRyYWx5dGljcy9wb3NlLnB5)	69.64%	[17 Missing ⚠️ ](https://app.codecov.io/gh/iterative/datachain/pull/613?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative)
> [src/datachain/model/ultralytics/segment.py](https://app.codecov.io/gh/iterative/datachain/pull/613?src=pr&el=tree&filepath=src%2Fdatachain%2Fmodel%2Fultralytics%2Fsegment.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative#diff-c3JjL2RhdGFjaGFpbi9tb2RlbC91bHRyYWx5dGljcy9zZWdtZW50LnB5)	55.26%	[17 Missing ⚠️ ](https://app.codecov.io/gh/iterative/datachain/pull/613?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=iterative)

These methods are tested in `examples` ([1](https://github.com/iterative/datachain/pull/613/files#diff-f78654382c7c2503e3f1de7016f0726f7874418090fe935e7c69843ffeb95a38), [2](https://github.com/iterative/datachain/pull/613/files#diff-02a09369bb8dc48e440a22324c9a7836976efc92d274db90ad90e7b51e320f23), [3](https://github.com/iterative/datachain/pull/613/files#diff-d01e3b43ca7c4f0ffc0431701315050cee018170af61f79eca070582a0b7fe3b)). Adding tests for them requires installing ultralytics dependency and downloading models, which will increase tests run time. Still not sure if this is a good idea. Also we can do it in follow-up PRs.

File: src/datachain/model/bbox.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    """
+    A data model for representing bounding box.
+
+    Attributes:
+        title (str): The title of the bounding box.
+        coords (list[int]): The coordinates of the bounding box.
+
+    The bounding box is defined by two points:
+        - (x1, y1): The top-left corner of the box.
+        - (x2, y2): The bottom-right corner of the box.
```
Comment (mattseddon) [inline src/datachain/model/bbox.py:16]: [Q] We currently support N `File` models on the same record as long as they have different names. Should these models have a file they are associated with?

Comment (dreadatour) [reply src/datachain/model/bbox.py:16]: Good question indeed. I am thinking about use cases. Do we have any examples of other signals with a file associated with? 🤔

Comment (mattseddon) [reply src/datachain/model/bbox.py:16]: The first one that comes to mind would be extracting images or text or tables from a PDF and having bounding boxes attached to the original for where they've been extracted. 

I'm not criticising or saying we should add this now but this is something to consider as we start adding these model types.

Comment (dreadatour) [reply src/datachain/model/bbox.py:16]: > I'm not criticising or saying we should add this now but this is something to consider as we start adding these model types.

I absolutely agree with you! This is very good point I wan't thinking about 👍

> The first one that comes to mind would be extracting images or text or tables from a PDF and having bounding boxes attached to the original for where they've been extracted.

I meant existing implementations or examples of signals for multiple files, was wondering may be you have any from the top of your head.

**Pull request merged**